File: 056 Merge Intervals/merge_intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
bool compareInterval(Interval i1, Interval i2) {
    return i1.start < i2.start;
}
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        sort(intervals.begin(), intervals.end(), compareInterval);
        int index = 0, left, right;
		vector<Interval> result;
		while (index < intervals.size()) {
			left = intervals[index].start;
			right = intervals[index].end;
			while (++index < intervals.size()) {
				if (intervals[index].start <= right) {
					right = max(right, intervals[index].end);
				}
				else break;
			}
			result.push_back(Interval(left, right));
		}
		return result;
    }
}; // Author: XC

File: 056 Merge Intervals/merge_intervals.cs
/**
 * Definition for an interval.
 * public class Interval {
 *     public int start;
 *     public int end;
 *     public Interval() { start = 0; end = 0; }
 *     public Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public IList<Interval> Merge(IList<Interval> intervals) {
        var sorted = intervals.OrderBy(x => x.start).ToList();
        var result = new List<Interval>();
        foreach (var i in sorted) {
            if (result.Count == 0) result.Add(i);
            else {
                int end = result.Last().end;
                if (i.start > end) result.Add(i);
                else result.Last().end = Math.Max(i.end, end);
            }
        }
        return result;
        
    }
} // Author: XC
